staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go

```diff
@@ -166,3 +166,59 @@ type TracingConfiguration struct {
 	// Defaults to 0.
 	SamplingRatePerMillion *int32
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityAndFairnessConfiguration holds API Priority and Fairness parameters.
+type PriorityAndFairnessConfiguration struct {
+	metav1.TypeMeta
+
+	// WorkEstimatorConfig holds work estimator parameters.
+	// +optional
+	WorkEstimatorConfiguration *WorkEstimatorConfiguration
+}
+
+// WorkEstimatorConfiguration holds work estimator parameters.
+type WorkEstimatorConfiguration struct {
+	// ListWorkEstimatorConfiguration holds work estimator parameters related to list requests.
+	// +optional
+	ListWorkEstimatorConfiguration *ListWorkEstimatorConfiguration
+	// MutatingWorkEstimatorConfiguration holds work estimator
+	// parameters related to watches of mutating objects.
+	// +optional
+	MutatingWorkEstimatorConfiguration *MutatingWorkEstimatorConfiguration
+
+	// MaximumSeats is the maximum number of seats a request can occupy
+	//
+	// NOTE: work_estimate_seats_samples metric uses the value of maximumSeats
+	// as the upper bound, so when we change maximumSeats we should also
+	// update the buckets of the metric.
+	// +optional
+	MaximumSeats uint64
+}
+
+// ListWorkEstimatorConfiguration holds work estimator parameters related to list requests.
+type ListWorkEstimatorConfiguration struct {
+	// ObjectsPerSeat is a number of objects from a single list request per seat.
+	// +optional
+	ObjectsPerSeat float64
+}
+
+// MutatingWorkEstimatorConfiguration holds work estimator
+// parameters related to watches of mutating objects.
+type MutatingWorkEstimatorConfiguration struct {
+	// Enabled enables mutating work estimator.
+	// This option will be removed once we have better proof from
```

Comment:
APIs don't remove values once introduced.  Also, we avoid booleans and use enumerated strings instead.

Are there not values for EventAdditionalDuration and WatchesPerSeat that effectively disable this already?

Reply:
We probably want to keep this as mentioned in https://github.com/kubernetes/kubernetes/pull/111222#discussion_r927347533. It is possible to "sort of" disable `mutatingWorkEstimator` by setting `WatchesPerSeat` to a very high number but I doubt it'd be convenient for a user. We could have a convention that when the value is set to `0` then the feature is disabled but I don't know if this is subtle enough.

I used `bool` here as I've seen it being used in other parts of the API, e.g: https://github.com/kubernetes/kubernetes/blob/71a58de48be918b779b8383d6427b603c5fc8401/pkg/apis/authorization/types.go#L135-L136

I understand that enums may be better overall but here we will always have exactly two options (enabled or disabled). Another choice could be to implement `EnabledFeatures` field which would be an array of `Feature`s where `Feature` is enumerated `string` that has values like `MutatingWorkEstimator`, `ListWorkEstimator`, etc.

If we can't remove values once introduced then this is the best option I'd see for that field. I agree it will be a pain to have this field present in the future when no longer needed. But at this moment it seems appropriate to code it like that. Let me know what you think about it.

I do have a side question though: since this is only covered by `v1alpha1` can't we really remove it later?

Reply:
Removed the `Enabled` field.

Reply:
It would be logical for `EventAdditionalDuration=0` to effectively disable this feature, but IIRC the code does not actually work that way --- the request width gets plumped out even if the extra duration is zero.